harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,471 @@
+
+                next_collection_ts = time.time() + self._collection_interval
+                self._collect_metrics()
+                wait_ts = max(next_collection_ts - time.time(), 0)
+                self._should_exit.wait(timeout=wait_ts)
+        finally:
+            self._metrics_queue.put(None)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+
+    def _collect_metrics(self) -> None:
+        for collector in self._metric_collectors:
+            timestamp = datetime.datetime.now(datetime.timezone.utc)
+            metrics = collector.collect()
+            if not metrics:
+                continue
+            group_metrics = _Metric(group=collector.group, metrics=metrics, timestamp=timestamp)
+            self._metrics_queue.put(group_metrics)
+
+
+class _Shipper(threading.Thread):
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        metrics_queue: queue.Queue,
+        flush_interval: int = 10,
+    ):
+        self._metrics_queue = metrics_queue
+        self._flush_interval = flush_interval
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._metrics: typing.List[_Metric] = []
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        """Start the thread and periodically ship metrics in queue to master.
+
+        This thread batches metrics and reports the batch every `flush_interval` seconds.
+        Currently we make one API call per metric reported, but in the future we should report each
+        batch with one API call.
+        """
+        while True:
+            metrics: typing.List[_Metric] = []
+            next_flush_ts = time.time() + self._flush_interval
+
+            # Pop messages from queue until the next scheduled flush time.
+            while True:
+                now = time.time()
+                timeout = next_flush_ts - now
+
+                if timeout <= 0:
+                    break
+
+                try:
+                    msg = self._metrics_queue.get(timeout=timeout)
+                    if msg is None:
+                        # Received shutdown message, ship accumulated metrics and exit.
+                        self._ship_metrics(metrics=metrics)
+                        return
+
+                    metrics.append(msg)
+                except queue.Empty:
+                    # Timeout reached, no messages in queue.
+                    break
+
+            self._ship_metrics(metrics=metrics)
+
+    def stop(self) -> None:
+        self._metrics_queue.put(None)
+
+    def _ship_metrics(self, metrics: typing.List[_Metric]) -> None:
+        for m in metrics:
+            # Append agent ID to every metric reported.
+            metric = {self._agent_id: m.metrics}
+            self._report_metrics(group=m.group, timestamp=m.timestamp, metrics=metric)
```

Comment:
Gonna see if I can keep this straight.

* `_Metric` is an object with `group`, `timestamp`, and `metrics`
* `metrics` is a list of `_Metric`s
* `metric` is a dict from `agent_id` to a `_Metric.metrics`
* `_Metric.metrics` is either a dict from `metric_name` to `metric_val` or a dict from a `device` to a dict from `metric_name` to `metric_val`

The word "metric" is doing a lot of work, here, and it's easy for me to lose track. At the cost of brevity, do you think the code would benefit from more differentiation?

Which definition of metric does `_report_metrics` operate on?

Reply:
What should I read to verify the correctness of the `metric` data structure?

Reply:
the metrics API expects an arbitrary dictionary. it's up to the backend to do whatever it needs to with that.

i have a comment in `_MetricGroupCollector` that documents the structure that profiler is reporting:

```
        Returns metrics as a dictionary mapping each metric name to its metric value.

        Certain metrics may be associated with additional labels (i.e. GPU UUIDs) in which case
        the returned dictionary should be nested with the label as keys. For example:
```

Reply:
>> What should I read to verify the correctness of the metric data structure?

> the metrics API expects an arbitrary dictionary. it's up to the backend to do whatever it needs to with that.

I think this is what I was after. This tells me that the format of the dict shipped to the backend isn't part of this review.

Please resolve if you agree, or send back to me if I should read into the format.

This, BTW, might be a good unit test.